File: src/infra/file-system/implementations/aws.file-system.ts
import {
  DeleteObjectCommand,
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from '@aws-sdk/client-s3';
import { FileAdapter } from '../adapters/file-adapter';

export class AwsFileSystem {
  private s3: S3Client;

  constructor() {
    this.s3 = new S3Client({
      region: process.env.AWS_DEFAULT_REGION,
      credentials: {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
      },
    });
  }

  async put(file: FileAdapter): Promise<void> {
    const params = new PutObjectCommand({
      Bucket: process.env.AWS_S3_BUCKET_NAME,
      Key: `${file.relativePath}/${file.name}`,
      Body: file.content,
      ContentType: file.contentType,
    });

    await this.s3.send(params);
  }

  async get(relativePath: string, fileName: string): Promise<FileAdapter> {
    const params = new GetObjectCommand({
      Bucket: process.env.AWS_S3_BUCKET_NAME,
      Key: `${relativePath}/${fileName}`,
    });

    const { Body, ContentType, ContentLength } = await this.s3.send(params);

    return {
      name: fileName,
      relativePath: relativePath,
      content: Buffer.from(await Body.transformToByteArray()),
      contentType: ContentType,
      length: ContentLength,
    };
  }

  async delete(key: string): Promise<void> {
    const params = new DeleteObjectCommand({
      Bucket: process.env.AWS_S3_BUCKET_NAME,
      Key: key,
    });

    await this.s3.send(params);
  }
}

File: src/infra/file_system/implementations/aws_file_system.py
import os
import boto3
from django.core.files.base import ContentFile
from src.infra.file_system.adapters.file_adapter import FileAdapter


class AwsFileSystem:
    s3 = boto3.client("s3")
    bucket = os.getenv("AWS_S3_BUCKET_NAME", "")

    @classmethod
    def put(self, file: FileAdapter):
        key = f"{file['relative_path']}/{file['name']}"
        self.s3.upload_fileobj(
            Fileobj=ContentFile(file["content"]),
            Bucket=self.bucket,
            Key=key,
        )

    @classmethod
    def get(self, relative_path: str, file_name: str):
        key = f"{relative_path}/{file_name}"

        s3_object = self.s3.get_object(Bucket=self.bucket, Key=key)
        content = s3_object["Body"]
        content_type = s3_object.get("ContentType")
        content_length = s3_object.get("ContentLength")

        return {
            "name": file_name,
            "relative_path": relative_path,
            "content": content,
            "content_type": content_type,
            "length": content_length,
        }

    @classmethod
    def delete(self, key: str):
        self.s3.delete_object(Bucket=self.bucket, Key=key)
